### PR TITLE
create a separate caml_obj_force_promote_to

### PR DESCRIPTION
**Pull request by sadiqj**
Just a quick PR to check this was in line with our earlier conversation @stedolan.

### CODE REVIEW

**Comment by stedolan** [comment]
Looks good!

For the sake of anyone following along: @sadiqj noticed that when Gc.promote_to is used to optimise parallel data structures, a good optimisation is to skip the promotion if there's only one thread running. However, since there are a couple of places where promotion is necessary for correctness, skipping it all the time seems dangerous, hence the split into two primitives.

I'll hold off on merging this just yet, though, as another change is needed for this to work: we do need to do the actual promotion if a second domain is started before the next GC. Rather than keeping track of promotions that have been elided, I think the best approach is just to do a minor GC before starting the second domain.

File: byterun/obj.c [hunk 1]
```diff
@@ -150,13 +150,34 @@ CAMLprim value caml_obj_compare_and_swap (value v, value f, value oldv, value ne
   return Val_int(res);
 }
 
-/* caml_promote_to(obj, upto) promotes obj to be as least as shared as upto */
+/* caml_promote_to(obj, upto) may promote obj to be as least as shared as upto.
+   Skips promotion if there is only a single domain. This should only be used in
+   situations where there is a performance need to 'front-load' the promotion
+   across domains, such as for queues. If you need to guarantee promotion to the
+   major heap for correctness, */
```
Comment (stedolan) [inline byterun/obj.c]: I think half of this sentence got lost somewhere!

File: byterun/obj.c [hunk 2]
```diff
@@ -150,13 +150,34 @@ CAMLprim value caml_obj_compare_and_swap (value v, value f, value oldv, value ne
...
 CAMLprim value caml_obj_promote_to (value obj, value upto)
 {
-  if (Is_block(upto) && Is_minor(upto)) {
+  if (Is_block(upto) && (Is_minor(upto) || caml_domain_alone()))
+  {
```
Comment (stedolan) [inline byterun/obj.c]: Indentation in the runtime is [K&R style](https://en.wikipedia.org/wiki/Indentation_style#K&R_style): function opening braces go on their own line, but other opening braces (for `if`, `while`, etc.) go on the same line as their statement.

File: byterun/obj.c (hunk 1 above)
Comment (sadiqj) [reply byterun/obj.c]: Whoops. I'm really not sure where that went. Fixed.

File: byterun/obj.c (hunk 2 above)
Comment (sadiqj) [reply byterun/obj.c]: Thanks. This should be resolved now.

**Pull request merged**